### DISCUSSION TITLE
Questions/guidance testing slurm support

### DISCUSSION THREAD

**Discussion by MEllis-github** [sky-shift/skyshift#92] [General]
I'm looking forward to taking the slurm support for a test drive! 

Questions
- [x] With respect to requirements, does the current support via the REST API and/or the future CLI support assume a container manager (docker / podman / ...) is installed?  
- [x] What would you recommend for an initial smoke test?
- [x] @dc-base are you the POC for these questions, are there others?

**Comment by michaelzhiluo** [comment]
The Slurm PR is in #4. Currently, Slurm PR supports Rest API; however, the CLI version is being implemented. Both ways to poll the Slurm Cluster assume that there is an exsiting container manager - this compatibility layer currently supports Docker, ContainerD, Singularity, and Podman. (correct me if wrong, @dc-base )

We haven't tested custom installation for Slurm on any machine, but Daron has an easy Slurm setup script: https://github.com/dc-base/slurm-auto-script

**Reply by michaelzhiluo** [reply]
Also @AlexCuadron will begin testing Slurm after the `skyctl exec` PR.

**Reply by daronjchang** [reply]
Current compatibility layer supports Docker, ContainerD, and Singularity. Podman and PodmanHPC are planned to be added before the PR is finalized. 

Current testing is done on locally using REST. 
CLI is still in progress and being tested on the Berkeley Watson cluster with no container manager, and locally on a smaller scale with the container managers.

**Comment by daronjchang** [comment]
The easiest way to do a smoke test is using the REST API on a local machine. Most package managers already compile slurm with the REST enabled, and an additional package with Slurmrestd just needs to be installed. 

For setting up Slurmrestd, I recommend a local single node configuration listening on a unix socket, then trying to curl the endpoint located at 'http://localhost:8080/slurm/{open_api_ver}/ping'
Using:
      curl --unix-socket {socket_path} 'http://localhost:8080/slurm/{open_api_ver}/ping'

The simplest smoke test is pulling the dockerhub hello-world image and checking the job output log.

**Comment by MEllis-github** [comment]
Good info., thanks all!

I have (user level) access to several slurm clusters, none of which have the REST plugin, but one at least has a container manager. I can stand-up a VM or a local configuration, but assumed that wouldn't be helpful, perhaps prematurely.

**Comment by daronjchang** [comment]
The local configuration can be used to test out the REST version. I expect
to push out CLI version by the end of the week.

**Reply by dmatch01** [reply]
@MEllis-github I suggest to wait for the CLI since it will take some time to get the right folks to install the REST plugin and other associated requirements.

**Reply by MEllis-github** [reply]
@dc-base any updates/recommendations on when to start testing the slurm CLI ? Thanks in advance!

**Comment by MEllis-github** [comment]
Thanks for the discussion on status and local smoke testing above!

Next questions are
- [x] Is documentation for slurm support (e.g. quickstart, examples, etc.) being created and if so where? I've been looking in the `slurm` `docs` and `main` branches so far, in README.md, docs/, and skyflow/cli respectively.
- [x] What is the first step to discover/attach a slurm cluster assuming it is remote and authenticated? The slurm cluster I am trying to attach is not listed in my `skyctl get clusters` and `skyctl create... --manager slurm... ` otherwise looks like it might trigger creation (versus attachment only).

```
$ skyctl create cluster --manager slurm --help
Usage: skyctl create cluster [OPTIONS] NAME

  Attaches a new cluster.

Options:
  --manager TEXT       Cluster manager type (e.g. k8, slurm).  [default: k8;
                       required]
  --cpus TEXT          Number of vCPUs per node (e.g. 1, 1+).
  --memory TEXT        Amount of memory each instance must have in GB. (e.g.
                       32, 32+).
  --disk_size INTEGER  OS disk size in GBs
  --accelerators TEXT  Type and number of GPU accelerators to use
  --ports TEXT         Ports to open on the cluster
  --num_nodes INTEGER  Number of SkyPilot nodes to allocate to the cluster
                       [default: 1]
  --cloud TEXT
  --region TEXT
  --attached           True if cluster is already created and needs to be
                       attached to Skyflow.
  --help               Show this message and exit.
```

**Reply by daronjchang** [reply]
I pushed an initial quick start to /examples/slurm-configuration/slurm_README.md

I will merge Michael's doc branch into mine and move it over soon.

**Reply by MEllis-github** [reply]
Awesome, thanks, I will check it out!

**Reply by MEllis-github** [reply]
@dc-base I created an issue here to double check some assumptions and debug an issue getting started: https://github.com/michaelzhiluo/skyflow/issues/113. Depending on the info., we can change it from a question-issue to a documentation-issue or whatever next step is appropriate. I can help move answers to documentation as appropriate.

CC: @dmatch01
